packages/babel-plugin-proposal-decorators/CONTRIB.md

```diff
@@ -0,0 +1,374 @@
+
+```js
+let Class;
+class _Class {
+```
+
+This class is being decorated directly, which means that the decorator may
+replace the class itself. Class bindings are not mutable, so we need to create a
+new `let` variable for the decorated class.
+
+
+```js
+  static {
+    let ret = applyDecs(
+      this,
+      elementDecs,
+      classDecs
+    );
+
+    initA = ret[0];
+    callB = ret[1];
+    initC = ret[2];
+    getC = ret[3];
+    setC = ret[4];
+    initInstance = ret[5];
+    Class = ret[6];
+    initClass = ret[7];
+  }
+```
+
+Next, we immediately define a `static` block which actually applies the
+decorators. This is important because we must apply the decorators _after_ the
+class prototype has been fully setup, but _before_ static fields are run, since
+static fields should only see the decorated version of the class.
+
+We apply the decorators to class elements and the class itself, and the
+application returns an array of values that are used to populate all of the
+local variables we defined earlier. The array's order is fully deterministic, so
+we can assign the values based on an index we can calculate ahead of time.
+
+We'll come back to `applyDecs` in a bit to dig into what its format is exactly,
+but now let's dig into the new definitions of our class elements.
+
+```js
+  a = (initInstance(this), initA(this, 123));
+```
+
+Alright, so previously this was a simple class field. Since it's the first field
+on the class, we've updated it to immediately call `initInstance` in its
+initializer. This calls any initializers added with `addInitializer` for all of
+the per-class values (methods and accessors), which should all be setup on the
+instance before class fields are assigned. Then, it calls `initA` to get the
+initial value of the field, which allows initializers returned from the
+decorator to intercept and decorate it. It's important that the initial value
+is used/defined _within_ the class body, because initializers can now refer to
+private class fields, e.g. `a = this.#b` is a valid field initializer and would
+become `a = initA(this, this.#b)`, which would also be valid. We cannot
+extract initializer code, or any other code, from the class body because of
+this.
+
+Overall, this decoration is pretty straightforward other than the fact that we
+have to reference `initA` externally.
+
+```js
+  static #b(...args) {
+    callB(this, args);
+  }
+  static x() {
+    console.log('foo');
+  }
+```
+
+Next up, we have a private static class method `#b`. This one is a bit more
+complex, as our definition has been broken out into 2 parts:
+
+1. `static #b`: This is the method itself, which being a private method we
+  cannot overwrite with `defineProperty`. We also can't convert it into a
+  private field because that would change its semantics (would make it
+  writable). So, we instead have it proxy to the locally scoped `callB`
+  variable, which will be populated with the fully decorated method.
```

Comment:
Should this code log `true` or `false`?

```ks
function x() {}

function replace() {
  return x;
}

class C {
  @replace
  static #m() {}
  
  static getM() {
    return this.#m;
  }
}

console.log(C.getM() === x);
```

if it should log `true`, then this proxy does not work correctly. We might decide that it's ok and it's a known limitation, but we could make it work by converting `#m` to a field and replacing `obj.#m = val` assignments with `obj, val, _readOnlyError("#m")` calls (`_readOnlyError` is an existing helper).

Reply:
This code should return true, but it definitely is not a common use case so it would be alright if we did not support this. I think the larger issue is that it seems private fields cannot be accessed prior to initialization, so for instance:

```js
let getA;

class Foo {
  #a = ((this.#b = (v) => v), this.#b(123));

  #b;

  getA() {
    return this.#a;
  }
}

(new Foo()).getA()
```

This code does not work. With a true private method, it _would_ work (tested in Chrome at least), and I think this is generally a more common pattern so I think we'll still want to do the proxy method here.

Reply:
Uhm right, private methods would need to be "hoisted" before all the private fields when converting them.

Reply:
hmm, yeah I think we can do that and that should work actually, will update